demo/src/package-lock-generator.ts

```diff
@@ -0,0 +1,238 @@
+const MEMORY_RESERVED_MB = 128;
+
+/**
+ * Basic representation of the contents of package-lock.json file.
+ */
+type PackageDefinition = {
+  dependencies?: {[key: string]: PackageDefinition};
+  version?: string;
+};
+
+/**
+ * Flat representation of a package-lock file.
+ */
+export type PackageVersionMap = {
+  [name: string]: string
+};
+
+/**
+ * Used to generate package locks for requested packages. These package locks
+ * consist of both dependencies and devDependencies. This is achieved by
+ * retrieving the packages package.json file and running `npm install` with that
+ * file.
+ *
+ * Response times vary by the number of transitive dependencies of a package.
+ * All generated package locks are persisted and reused once it has initially
+ * been generated. An in-memory cache also exists to reduce the number of reads.
+ * For example, when loading a module based demo, hundreds of requests can be
+ * generated reading the same package lock. However, responses are cached in an
+ *
+ * This also supports concurrent requests to generate the same package lock
+ * through deduplication of requests.
+ */
+export class PackageLockGenerator {
+  // Instantiate a cache which dynamically determines cache size based on how
+  // much memory is being used.
+  private cache = new Cache<Buffer>((_currentSize) => {
+    const currentHeapBytes = process.memoryUsage().heapUsed;
+    const currentHeapMB = currentHeapBytes / 1024 / 1024;
+    return MEMORY_TOTAL_MB - currentHeapMB > MEMORY_RESERVED_MB;
+  });
+  // Map to dedupe pending installs if there are concurrent requests.
+  private pendingInstalls = new Map<string, Promise<PackageVersionMap|null>>();
+  // Persistent database to store generated package locks.
+  private firestore?: Firestore;
+
+  constructor() {
+    if (process.env.NODE_ENV === 'production') {
+      firebaseAdmin.initializeApp(
+          {credential: firebaseAdmin.credential.applicationDefault()});
+
+      this.firestore = firebaseAdmin.firestore();
+      this.firestore.settings({timestampsInSnapshots: true});
+    } else {
+      console.log('Running without Firestore backing.');
+    }
+  }
+
+  /**
+   * Fetch the package versions for the specified package (in the form
+   * @scope/package@1.0.0). This is done by checking the in-memory cache, then
+   * Firestore. If it does not exist, it is generated.
+   */
+  async get(packageString: string): Promise<PackageVersionMap|null> {
+    const valueFromCache = this.cache.get(packageString);
+    if (valueFromCache) {
+      return await this.uncompressData(valueFromCache);
+    }
+
+    let firestoreDocId;
+
+    // Check if there is already a persisted value in datastore.
+    if (this.firestore) {
+      firestoreDocId = this.firestore.collection('package-locks')
+                           .doc(packageString.replace('/', '__'))
+      const doc = await firestoreDocId.get();
+      if (doc.exists) {
+        const packageVersionMap = doc.data() as PackageVersionMap;
```

Comment:
Possible todo: have a pending value for the doc, and if the doc is pending, wait on it being written so that it's more likely that only one process writes to firestore.

Reply:
Yeah, that's a possible future improvement, but seems like a minor optimisation at this stage and correctness isn't an issue here.